Add ZSA functionalities

ConstanceBeguier opened the pull request:
The following modifications are required to support the ZSA circuit changes:
- Added a function CommitDomain::new_with_separate_domains to initialize the CommitDomain from two different personalizations. This is required to be able to use an identical R for ZEC and ZSA note commitments but a different Q.

ConstanceBeguier committed:
Add ZSA functionality

The following modifications are required to support the ZSA circuit changes:
- Added a function CommitDomain::new_with_personalization to initialize the CommitDomain from two different personalizations. This is required to be able to use an identical R for ZEC and ZSA note commitments but a different Q.
- Added a function commit_from_hash_point to evaluate a Sinsemilla commit from a hash point.
- Added the ability to hash and blind independently.

str4d commented on src/lib.rs:
```diff
@@ -319,4 +352,27 @@ mod tests {
+        // Evaluate the commitment with the following steps
+        // 1. hash msg
+        // 2. evaluate the commitment from the hash point
+        let hash_point = domain.M.hash_to_point(msg.into_iter());
```
This should be testing the public API:
```suggestion
        let hash_point = domain.hash_to_point(msg.into_iter());
```

ConstanceBeguier replied:
Done

str4d commented on src/lib.rs:
```diff
@@ -212,6 +212,17 @@ impl CommitDomain {
         }
     }
 
+    /// Constructs a new `CommitDomain` from different values for `hash_domain` and `blind_domain`
```
```suggestion
    /// Constructs a new `CommitDomain` from different values for `hash_domain` and `blind_domain`.
```

This needs documentation indicating how this is intended to be used (ideally with a specification link, e.g. to how it is used in ZSAs).

ConstanceBeguier replied:
Done

str4d commented on src/lib.rs:
```diff
@@ -212,6 +212,17 @@ impl CommitDomain {
 
+    /// Constructs a new `CommitDomain` from different values for `hash_domain` and `blind_domain`
+    pub fn new_with_personalization(hash_domain: &str, blind_domain: &str) -> Self {
+        let m_prefix = format!("{}-M", hash_domain);
```
Instead of implementing this by duplication, change `CommitDomain::new` to call this constructor setting both arguments to `domain`.

ConstanceBeguier replied:
Done

str4d commented on src/lib.rs:
```diff
@@ -212,6 +212,17 @@ impl CommitDomain {
     }
 
+    /// Constructs a new `CommitDomain` from different values for `hash_domain` and `blind_domain`
+    pub fn new_with_personalization(hash_domain: &str, blind_domain: &str) -> Self {
```
```suggestion
    pub fn new_with_separate_domains(hash_domain: &str, blind_domain: &str) -> Self {
```
(`CommitDomain::new` is always personalized.)

ConstanceBeguier replied:
Done

str4d commented on src/lib.rs:
```diff
@@ -226,6 +237,26 @@ impl CommitDomain {
+    /// $\mathsf{SinsemillaHashToPoint}$ from [§ 5.4.1.9][concretesinsemillahash].
+    ///
+    /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+    pub fn hash_to_point(&self, msg: impl Iterator<Item = bool>) -> CtOption<pallas::Point> {
```
This needs documentation explaining when and why it would be used. Having a `hash_to_point` method on a `CommitDomain` explicitly breaks the type safety we were previously providing with these structs, so I want the caller to be very aware they are doing something cryptographically "interesting".

Maybe also rename to make it explicit to downstream reviewers:
```suggestion
    pub fn low_level_hash_to_point(&self, msg: impl Iterator<Item = bool>) -> CtOption<pallas::Point> {
```

Also, given my comment on `Self::commit_from_hash_point` and depending on what your intention for it is, it may be better to return a struct from this method that contains _both_ the hash-to-point intermediate, and the final commitment (which would keep the domain consistency while avoiding any performance problem):
```rust
pub struct CommitmentWithIntermediate {
    intermediate: pallas::Point,
    commitment: pallas::Point,
}

impl CommitmentWithIntermediate {
    pub fn low_level_intermediate(&self) -> pallas::Point {
        self.intermediate
    }

    pub fn commitment(&self) -> pallas::Point {
        self.commitment
    }
}
```

(with appropriate documentation), and then change this function to:
```suggestion
    pub fn low_level_commit(&self, msg: impl Iterator<Item = bool>) -> CtOption<CommitmentWithIntermediate> {
```

ConstanceBeguier replied:
I agree that the hash_to_point and commit_from_hash_point functions are not essential and potentially unsafe.
So I removed them.

str4d commented on src/lib.rs:
```diff
@@ -226,6 +237,26 @@ impl CommitDomain {
+    /// where `SinsemillaHash(personalization, msg) = hash_point`
+    /// and `R` is derived from the `personalization`.
+    #[allow(non_snake_case)]
+    pub fn commit_from_hash_point(
```
Why does this need to be exposed? The caller could equivalently clone their `msg` and provide it to both `Self::commit` and `Self::low_level_hash_to_point`. It would also be safer, as it wouldn't be exposing a capability to derive a commitment from the `-r` domain that is inconsistent with the hash point's `-M` domain.

The only reason I can think of to expose this would be performance, and I think any performance concerns here are negligible (or should be backed up with benchmarks showing why the degradation is unacceptable).

ConstanceBeguier replied:
I agree that the hash_to_point and commit_from_hash_point functions are not essential and potentially unsafe.
So I removed them.

str4d commented on src/lib.rs:
```diff
@@ -319,4 +352,27 @@ mod tests {
+        let commit2 = domain.commit_from_hash_point(hash_point, &rcm);
+
+        // Test equality
+        assert_eq!(commit1.unwrap(), commit2.unwrap());
```
If the above API changes are made, then extend this test to additionally verify the intermediate.

ConstanceBeguier replied:
This test has been removed because the hash_to_point and commit_from_hash_point functions have been removed

str4d reviewed (changes requested):
Reviewed 3542e897d42187878ff9da714f3221cc1774caa7 with @daira.

ConstanceBeguier committed:
Take into account review of commit 3542e89

- Rename new_with_personalization by new_with_separate_domains
- Remove hash_to_point and commit_from_hash_point functions that are not essential

daira commented on src/lib.rs:
```diff
@@ -203,8 +203,17 @@ pub struct CommitDomain {
+        Self::new_with_separate_domains(domain, domain)
+    }
+
+    /// Constructs a new `CommitDomain` from different values for `hash_domain` and `blind_domain`
```
```suggestion
    /// Constructs a new `CommitDomain` from separate values for `hash_domain` and `blind_domain`.
    ///
```

str4d reviewed (approved):
ACK fa55f0bde196e2f884871a4dc743bede1f433f33

daira reviewed (approved):
utACK (comment is nonblocking).
